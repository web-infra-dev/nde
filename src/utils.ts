import os from 'node:os';
import path from 'node:path';
import { type NodeFileTraceOptions, nodeFileTrace } from '@vercel/nft';
import createDebug from 'debug';
import fse from 'fs-extra';
import { parseNodeModulePath } from 'mlly';
import type { PackageJson } from 'pkg-types';

export const debug = createDebug('nde');

export type TracedPackage = {
  name: string;
  versions: Record<
    string,
    {
      pkgJSON: PackageJson;
      path: string;
      isDirectDep: boolean;
      files: string[];
    }
  >;
};

export type TracedFile = {
  path: string;
  subpath: string;
  parents: string[];
  isDirectDep: boolean;

  pkgPath: string;
  pkgName: string;
  pkgVersion?: string;
};

function applyPublicCondition(pkg: PackageJson) {
  if (pkg?.publishConfig?.exports) {
    pkg.exports = pkg?.publishConfig?.exports;
  }
}

interface WritePackageOptions {
  pkg: TracedPackage;
  version: string;
  projectDir: string;
  _pkgPath?: string;
}

export const writePackage = async (options: WritePackageOptions) => {
  const { pkg, version, projectDir, _pkgPath } = options;
  const pkgPath = _pkgPath || pkg.name;
  for (const src of pkg.versions[version].files) {
    if (src.includes('node_modules')) {
      const { subpath } = parseNodeModulePath(src);
      // biome-ignore lint/style/noNonNullAssertion: <explanation>
      const dest = path.join(projectDir, 'node_modules', pkgPath, subpath!);
      const dirname = path.dirname(dest);
      await fse.ensureDir(dirname);
      await fse.copyFile(src, dest);
    } else {
      // workspace package
      const subpath = path.relative(pkg.versions[version].path, src);
      const dest = path.join(projectDir, 'node_modules', pkgPath, subpath);
      const dirname = path.dirname(dest);
      await fse.ensureDir(dirname);
      await fse.copyFile(src, dest);
    }
  }

  const { pkgJSON } = pkg.versions[version];
  applyPublicCondition(pkgJSON);

  const packageJsonPath = path.join(
    projectDir,
    'node_modules',
    pkgPath,
    'package.json',
  );
  await fse.ensureDir(path.dirname(packageJsonPath));
  await fse.writeFile(packageJsonPath, JSON.stringify(pkgJSON, null, 2));
};

const isWindows = os.platform() === 'win32';
export const linkPackage = async (
  from: string,
  to: string,
  projectRootDir: string,
) => {
  const src = path.join(projectRootDir, 'node_modules', from);
  const dest = path.join(projectRootDir, 'node_modules', to);
  const dstStat = await fse.lstat(dest).catch(() => null);
  const exists = dstStat?.isSymbolicLink();

  if (exists) {
    return;
  }
  await fse.mkdir(path.dirname(dest), { recursive: true });
  await fse
    .symlink(
      path.relative(path.dirname(dest), src),
      dest,
      isWindows ? 'junction' : 'dir',
    )
    .catch(error => {
      console.error('Cannot link', from, 'to', to, error);
    });
};

interface ReadDirOptions {
  filter?: (filePath: string) => boolean;
}

export const readDirRecursive = async (
  dir: string,
  options: ReadDirOptions = {},
): Promise<string[]> => {
  const { filter } = options;
  const files = await fse.readdir(dir, { withFileTypes: true });
  const filesAndDirs = await Promise.all(
    files.map(async file => {
      const resolvedPath = path.resolve(dir, file.name);
      if (file.isDirectory()) {
        return readDirRecursive(resolvedPath, options);
      }
      return filter && !filter(resolvedPath) ? [] : resolvedPath;
    }),
  );
  return filesAndDirs.flat();
};

export const isFile = async (file: string) => {
  try {
    const stat = await fse.stat(file);
    return stat.isFile();
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  } catch (error: any) {
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
};

export const findEntryFiles = async (
  rootDir: string,
  entryFilter?: (filePath: string) => boolean,
) => {
  const files = await readDirRecursive(rootDir, { filter: entryFilter });
  return files.filter(
    file =>
      file.endsWith('.mjs') || file.endsWith('.cjs') || file.endsWith('.js'),
  );
};

export const findPackageParents = (
  pkg: TracedPackage,
  version: string,
  tracedFiles: Record<string, TracedFile>,
) => {
  const versionFiles: TracedFile[] = pkg.versions[version].files.map(
    path => tracedFiles[path],
  );

  const parentPkgs = [
    ...new Set(
      versionFiles.flatMap(file =>
        // Because it supports copyWholePackage configuration, not all files exist.
        file?.parents
          .map(parentPath => {
            const parentFile = tracedFiles[parentPath];

            // when parent does not exist, parent may be an entry file.
            if (!parentFile || parentFile.pkgName === pkg.name) {
              return null;
            }
            return `${parentFile.pkgName}@${parentFile.pkgVersion}`;
          })
          .filter(Boolean),
      ),
    ),
  ];
  return parentPkgs.filter(parentPkg => parentPkg) as string[];
};

async function serializeMap(map: Map<string, unknown>): Promise<string> {
  const resolvedMap = new Map<string, unknown>();

  // Resolve all promises in the map
  await Promise.all(
    Array.from(map.entries()).map(async ([key, value]) => {
      resolvedMap.set(
        key,
        value instanceof Promise ? await Promise.resolve(value) : value,
      );
    }),
  );

  return JSON.stringify(resolvedMap, (key, value) => {
    if (value === null) {
      return undefined;
    }
    if (value instanceof Map) {
      return {
        dataType: 'Map',
        value: Array.from(value.entries()),
      };
    }
    if (value instanceof Set) {
      return {
        dataType: 'Set',
        value: Array.from(value),
      };
    }
    return value;
  });
}

// Function to deserialize a Map with Set values
function deserializeMap(serializedData: string) {
  return JSON.parse(serializedData, (key, value) => {
    if (value && value.dataType === 'Map') {
      return new Map(value.value);
    }
    if (value && value.dataType === 'Set') {
      return new Set(value.value);
    }
    return value;
  });
}

export interface CacheOptions {
  fileCache: boolean;
  analysisCache: boolean;
  symlinkCache: boolean;
  cacheDir: string;
}

const loadCache = async (filePath: string, enabled: boolean) => {
  if (enabled && (await fse.pathExists(filePath))) {
    debug('load cache:', filePath);
    const data = (await fse.readFile(filePath)).toString();
    return deserializeMap(data);
  }
  return undefined;
};

const writeCache = async (filePath: string, cacheMap: Map<string, unknown>) => {
  const newCacheMap = new Map();
  for (const [key, value] of cacheMap.entries()) {
    if (key.includes('node_modules/')) {
      newCacheMap.set(key, value);
    }
  }
  await fse.writeFile(filePath, await serializeMap(newCacheMap));
  console.log(`write ${path.basename(filePath)} finish`);
};

export const traceFiles = async ({
  entryFiles,
  sourceDir,
  base = '/',
  cacheOptions,
  traceOptions,
}: {
  entryFiles: string[];
  sourceDir: string;
  base?: string;
  cacheOptions: CacheOptions;
  traceOptions?: NodeFileTraceOptions;
}) => {
  const { cacheDir, fileCache, analysisCache, symlinkCache } = cacheOptions;
  const analysisCacheFile = path.join(cacheDir, 'analysis-cache.json');
  const fileCacheFile = path.join(cacheDir, 'file-cache.json');
  const symlinkCacheFile = path.join(cacheDir, 'symlink-cache.json');

  const cache = {
    analysisCache: await loadCache(analysisCacheFile, analysisCache),
    fileCache: await loadCache(fileCacheFile, fileCache),
    symlinkCache: await loadCache(symlinkCacheFile, symlinkCache),
  };

  const res = await nodeFileTrace(entryFiles, {
    base,
    processCwd: sourceDir,
    cache,
    ...traceOptions,
  });

  if (analysisCache || fileCache || symlinkCache) {
    await fse.ensureDir(cacheDir);
    if (cache.analysisCache && analysisCache) {
      writeCache(analysisCacheFile, cache.analysisCache);
    }
    if (cache.fileCache && fileCache) {
      writeCache(fileCacheFile, cache.fileCache);
    }
    if (cache.symlinkCache && symlinkCache) {
      writeCache(symlinkCacheFile, cache.symlinkCache);
    }
  }

  return res;
};

export const resolveTracedPath = async (
  base: string,
  p: string,
): Promise<string> => fse.realpath(path.resolve(base, p));

export const isSubPath = (parentPath: string, childPath: string) => {
  if (!parentPath || !childPath) {
    return false;
  }

  const relative = path.relative(parentPath, childPath);
  return relative && !relative.startsWith('..');
};
