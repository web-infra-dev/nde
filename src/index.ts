import path from "node:path";
import type { NodeFileTraceOptions } from "@vercel/nft";
import fse from "fs-extra";
import { parseNodeModulePath } from "mlly";
import type { PackageJson } from "pkg-types";
import { readPackageJSON } from "pkg-types";
import pkgUp from "pkg-up";
import semver from "semver";

import {
	type CacheOptions,
	type TracedFile,
	type TracedPackage,
	traceFiles as defaultTraceFiles,
	findEntryFiles,
	findPackageParents,
	isFile,
	isSubPath,
	linkPackage,
	readDirRecursive,
	resolveTracedPath,
	writePackage,
} from "./utils";
import { debug } from "./utils";

export type { NodeFileTraceOptions } from "@vercel/nft";
export { nodeFileTrace } from "@vercel/nft";

export const nodeDepEmit = async ({
	appDir,
	sourceDir,
	includeEntries,
	traceFiles = defaultTraceFiles,
	entryFilter,
	modifyPackageJson,
	copyWholePackage,
	cacheOptions = {
		cacheDir: ".modern-js/deploy",
		analysisCache: false,
		fileCache: false,
		symlinkCache: false,
	},
	traceOptions,
}: {
	/**
	 * Directory of the project
	 */
	appDir: string;
	/**
	 * The directory where the code will be analyzed, all js files in that directory will be used as entries, and the node_modules directory will be generated in that directory
	 */
	sourceDir: string;
	/**
	 * Some files to include, generally some files not analyzed by the code
	 */
	includeEntries?: string[];
	traceFiles?: typeof defaultTraceFiles;
	entryFilter?: (filePath: string) => boolean;
	modifyPackageJson?: (pkgJson: PackageJson) => PackageJson;
	copyWholePackage?: (pkgName: string, pkgJSON: PackageJson) => boolean;
	cacheOptions?: CacheOptions;
	traceOptions?: NodeFileTraceOptions;
}) => {
	const base = "/";
	const entryFiles = await findEntryFiles(sourceDir, entryFilter);

	debug("trace files start");
	const fileTrace = await traceFiles({
		entryFiles: entryFiles.concat(includeEntries || []),
		sourceDir,
		cacheOptions: {
			...cacheOptions,
			cacheDir: path.resolve(appDir, cacheOptions.cacheDir),
		},
		base,
		traceOptions,
	});
	debug("trace files end");
	const currentProjectModules = path.join(appDir, "node_modules");
	// Because vercel/nft may find inaccurately, we limit the range of query of dependencies
	const dependencySearchRoot = path.resolve(appDir, "../../../../../../");

	const tracedFiles: Record<string, TracedFile> = Object.fromEntries(
		(await Promise.all(
			[...fileTrace.reasons.entries()].map(async ([_path, reasons]) => {
				if (reasons.ignored) {
					return;
				}
				const filePath = await resolveTracedPath(base, _path);

				if (
					isSubPath(sourceDir, filePath) ||
					(isSubPath(appDir, filePath) &&
						!isSubPath(currentProjectModules, filePath))
				) {
					return;
				}

				if (!(await isFile(filePath))) {
					return;
				}

				let baseDir: string | undefined;
				let pkgName: string | undefined;
				let subpath: string | undefined;
				let pkgPath: string | undefined;

				if (filePath.includes("node_modules")) {
					const parsed = parseNodeModulePath(filePath);
					baseDir = parsed.dir;
					pkgName = parsed.name;
					subpath = parsed.subpath;
					// biome-ignore lint/style/noNonNullAssertion: <explanation>
					pkgPath = path.join(baseDir!, pkgName!);
				} else {
					// For @modern-js/utils, since there are some pre-bundled packages in the package that have their own package.json,
					// and since the relationship between these files uses relative paths, some special handling is required
					const MODERN_UTILS_PATH = "packages/toolkit/utils";
					const MODERN_UTILS_PATH_REGEX = new RegExp(
						`(.*${MODERN_UTILS_PATH})`,
					);
					const match = filePath.match(MODERN_UTILS_PATH_REGEX);

					const packageJsonPath: string | null = match
						? path.join(match[0], "package.json")
						: await pkgUp({ cwd: path.dirname(filePath) });

					if (
						packageJsonPath &&
						isSubPath(dependencySearchRoot, packageJsonPath)
					) {
						const packageJson: PackageJson =
							await fse.readJSON(packageJsonPath);

						pkgPath = baseDir = path.dirname(packageJsonPath);
						subpath = path.relative(baseDir, filePath);
						pkgName = packageJson.name;
					}
				}

				if (!baseDir) {
					return;
				}

				const parents = await Promise.all(
					[...reasons.parents].map((p) => resolveTracedPath(base, p)),
				);
				const tracedFile = {
					path: filePath,
					parents,
					isDirectDep: parents.some((parent) => {
						return (
							isSubPath(appDir, parent) &&
							!isSubPath(currentProjectModules, parent)
						);
					}),

					subpath,
					pkgName,
					pkgPath,
				} as TracedFile;

				return [filePath, tracedFile];
			}),
		).then((r) => r.filter(Boolean))) as [string, TracedFile][],
	);

	const tracedPackages: Record<string, TracedPackage> = {};
	for (const tracedFile of Object.values(tracedFiles)) {
		const { pkgName } = tracedFile;
		let tracedPackage = tracedPackages[pkgName];

		let pkgJSON = await readPackageJSON(tracedFile.pkgPath, {
			cache: true,
		}).catch(() => {});
		if (!pkgJSON) {
			pkgJSON = { name: pkgName, version: "0.0.0" } as PackageJson;
		}
		if (!tracedPackage) {
			tracedPackage = {
				name: pkgName,
				versions: {},
			};
			tracedPackages[pkgName] = tracedPackage;
		}

		// biome-ignore lint/style/noNonNullAssertion: <explanation>
		let tracedPackageVersion = tracedPackage.versions[pkgJSON.version!];
		if (!tracedPackageVersion) {
			tracedPackageVersion = {
				path: tracedFile.pkgPath,
				files: [],
				isDirectDep: false,
				pkgJSON,
			};
			if (tracedFile.isDirectDep) {
				tracedPackageVersion.isDirectDep = tracedFile.isDirectDep;
			}
			// biome-ignore lint/style/noNonNullAssertion: <explanation>
			tracedPackage.versions[pkgJSON.version!] = tracedPackageVersion;
		}

		tracedFile.pkgName = pkgName;
		tracedFile.pkgVersion = pkgJSON.version;

		const shouldCopyWholePackage = copyWholePackage?.(pkgName, pkgJSON);
		if (
			tracedFile.path.startsWith(tracedFile.pkgPath) &&
			// Merged package files are based on the version, not on paths, to handle some boundary cases
			tracedPackageVersion.pkgJSON.version === tracedFile.pkgVersion
		) {
			if (shouldCopyWholePackage) {
				const allFiles = await readDirRecursive(tracedFile.pkgPath);
				tracedPackageVersion.files.push(...allFiles);
			} else {
				tracedPackageVersion.files.push(tracedFile.path);
			}
		}
	}

	const multiVersionPkgs: Record<string, { [version: string]: string[] }> = {};
	const singleVersionPackages: string[] = [];
	for (const tracedPackage of Object.values(tracedPackages)) {
		const versions = Object.keys(tracedPackage.versions);
		if (versions.length === 1) {
			singleVersionPackages.push(tracedPackage.name);
			continue;
		}
		multiVersionPkgs[tracedPackage.name] = {};
		for (const version of versions) {
			// biome-ignore lint/style/noNonNullAssertion: <explanation>
			multiVersionPkgs[tracedPackage.name!][version!] = findPackageParents(
				tracedPackage,
				version,
				tracedFiles,
			);
		}
	}

	await Promise.all(
		singleVersionPackages.map((pkgName) => {
			const pkg = tracedPackages[pkgName];
			const version = Object.keys(pkg.versions)[0];
			return writePackage({
				pkg,
				version,
				projectDir: sourceDir,
			});
		}),
	);

	const projectPkgJson = await readPackageJSON(sourceDir).catch(
		() => ({}) as PackageJson,
	);

	for (const [pkgName, pkgVersions] of Object.entries(multiVersionPkgs)) {
		const versionEntires = Object.entries(pkgVersions).sort(
			([v1, p1], [v2, p2]) => {
				const shouldHoist1 =
					tracedPackages[pkgName]?.versions?.[v1]?.isDirectDep;
				const shouldHoist2 =
					tracedPackages[pkgName]?.versions?.[v2]?.isDirectDep;

				if (shouldHoist1 && !shouldHoist2) {
					return -1;
				}
				if (!shouldHoist1 && shouldHoist2) {
					return 1;
				}
				if (p1.length === 0) {
					return -1;
				}
				if (p2.length === 0) {
					return 1;
				}

				return semver.lt(v1, v2, { loose: true }) ? 1 : -1;
			},
		);

		for (const [version, parentPkgs] of versionEntires) {
			const pkg = tracedPackages[pkgName];

			const pkgDestPath = `.ndepe/${pkgName}@${version}/node_modules/${pkgName}`;
			await writePackage({
				pkg,
				version,
				projectDir: sourceDir,
				_pkgPath: pkgDestPath,
			});
			await linkPackage(pkgDestPath, `${pkgName}`, sourceDir);

			for (const parentPkg of parentPkgs) {
				const parentPkgName = parentPkg.replace(/@[^@]+$/, "");
				await (multiVersionPkgs[parentPkgName]
					? linkPackage(
							pkgDestPath,
							`.ndepe/${parentPkg}/node_modules/${pkgName}`,
							sourceDir,
						)
					: linkPackage(
							pkgDestPath,
							`${parentPkgName}/node_modules/${pkgName}`,
							sourceDir,
						));
			}
		}
	}

	const outputPkgPath = path.join(sourceDir, "package.json");

	const newPkgJson = {
		name: `${projectPkgJson.name || "modernjs-project"}-prod`,
		version: projectPkgJson.version || "0.0.0",
		private: true,
		dependencies: Object.fromEntries(
			[
				...Object.values(tracedPackages).map((pkg) => [
					pkg.name,
					Object.keys(pkg.versions)[0],
				]),
			].sort(([a], [b]) => a.localeCompare(b)),
		),
	};

	const finalPkgJson = modifyPackageJson?.(newPkgJson) || newPkgJson;

	await fse.writeJSON(outputPkgPath, finalPkgJson);
	debug("nodeDepEmit finish");
};
