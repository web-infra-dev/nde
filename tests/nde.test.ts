
import { expect, describe, it, afterEach } from 'vitest'
import path from 'node:path'
import fse from 'fs-extra'
import { handleDependencies } from '../src';

describe('handle dependencies', () => {
  const project1Dir = path.join(__dirname, 'fixtures/project1');
  const srcDir = path.join(project1Dir, 'src');
  const outputNodeModulesDir = path.join(srcDir, 'node_modules');
  const ndeDir = path.join(outputNodeModulesDir, '.nde');
  const outputPkgPath = path.join(srcDir, 'package.json');
  afterEach(async () => {
    await fse.remove(outputNodeModulesDir);
    await fse.remove(outputPkgPath);
  })
  it('basic usage', async() => {
    await handleDependencies({
      appDir: project1Dir,
      sourceDir: srcDir,
    })
    const items = await fse.readdir(ndeDir)
    expect(items).toMatchObject([ 'depd@1.1.2', 'depd@2.0.0' ]);
    const pkgJson = await fse.readJSON(outputPkgPath)
    expect(pkgJson).toMatchSnapshot();
  })
})


