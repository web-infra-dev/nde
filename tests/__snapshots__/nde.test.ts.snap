// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handle dependencies > basic usage 1`] = `
{
  "dependencies": {
    "accepts": "1.3.8",
    "cache-content-type": "1.0.1",
    "co": "4.6.0",
    "content-disposition": "0.5.4",
    "content-type": "1.0.5",
    "cookies": "0.9.1",
    "debug": "4.3.6",
    "deep-equal": "1.0.1",
    "delegates": "1.0.0",
    "depd": "2.0.0",
    "destroy": "1.2.0",
    "ee-first": "1.1.1",
    "encodeurl": "1.0.2",
    "escape-html": "1.0.3",
    "fresh": "0.5.2",
    "has-symbols": "1.0.3",
    "has-tostringtag": "1.0.2",
    "http-assert": "1.5.0",
    "http-errors": "1.8.1",
    "inherits": "2.0.4",
    "is-generator-function": "1.0.10",
    "keygrip": "1.1.0",
    "koa": "2.15.3",
    "koa-compose": "4.1.0",
    "koa-convert": "2.0.0",
    "media-typer": "0.3.0",
    "mime-db": "1.52.0",
    "mime-types": "2.1.35",
    "ms": "2.1.2",
    "negotiator": "0.6.3",
    "on-finished": "2.4.1",
    "only": "0.0.2",
    "parseurl": "1.3.3",
    "safe-buffer": "5.2.1",
    "setprototypeof": "1.2.0",
    "statuses": "1.5.0",
    "toidentifier": "1.0.1",
    "tsscmp": "1.0.6",
    "type-is": "1.6.18",
    "vary": "1.1.2",
    "ylru": "1.4.0",
  },
  "name": "project1-prod",
  "private": true,
  "version": "1.0.0",
}
`;
